ipc/provider/src/checkpoint.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: MIT
+//! Bottom up checkpoint manager
+
+use crate::config::Subnet;
+use crate::manager::{BottomUpCheckpointRelayer, EthSubnetManager};
+use anyhow::{anyhow, Result};
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use ipc_identity::{EthKeyAddress, PersistentKeyStore};
+use std::fmt::{Display, Formatter};
+use std::sync::{Arc, RwLock};
+use std::time::Duration;
+
+/// Tracks the metadata required for both top down and bottom up checkpoint submissions, such as
+/// parent/child subnet and checkpoint period.
+pub struct CheckpointMetadata {
+    parent: Subnet,
+    child: Subnet,
+    period: ChainEpoch,
+}
+
+pub struct BottomUpCheckpointManager<P, C> {
+    metadata: CheckpointMetadata,
+    parent_handler: P,
+    child_handler: C,
+}
+
+impl<P: BottomUpCheckpointRelayer, C: BottomUpCheckpointRelayer> BottomUpCheckpointManager<P, C> {
+    pub async fn new(
+        parent: Subnet,
+        child: Subnet,
+        parent_handler: P,
```

Comment:
What you could maybe do here is to initialize an `IpcProvider` with access to the parent and the child and that implements the operations that you need for the relayer

Reply:
I think that works too, just I feel that way we are making both `IpcProvider` and `BottomUpCheckpoint` quite messy. `IpcProvider` is like a wrapper that coordinates and exposes `SubnetManager`, it just handles the connection and calls into `SubnetManager`. If we shift Checkpoint implementation to it, it will be more complex than other methods. We can ask `IpcProvider` to call into `BottomUpCheckpointManager` if needed. That way I think it's neater.

Reply:
As discussed sync, let's keep the `IpcProvider` as a simple wrapper for external functions for now, and let's evolve it from user and DX feedback.